codestyle.cfg

```diff
@@ -0,0 +1,248 @@
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Configuration options for Mbed TLS code style
+
+# Line length options
+
+# Limit lines to 80 characters (or try to)
+code_width = 80
```

Comment:
There are times when going beyond 80 characters can be useful, as long as people don't always have very long lines.

80 characters is literally a limit from the 1980s, back when we used 13" CRTs and some linkers limited identifiers to 6 characters.

These days, screens are much wider, and it's easy to have two windows with > 80-character lines next to each other, so the limit is readability rather than technology.

I personally like line-length limits such as "try to keep to 80 chars, but hard-limit at 95 or 100 chars" (depending on project).

In this PR, early on we have

```
void Hacl_Curve25519_crypto_scalarmult(uint8_t *mypublic, uint8_t *secret, uint8_t *basepoint);
```

being replaced by

```
void Hacl_Curve25519_crypto_scalarmult(uint8_t *mypublic,
                                       uint8_t *secret,
                                       uint8_t *basepoint);
```

```
#  define KRML_HOST_SNPRINTF(buf, sz, fmt, arg) _snprintf_s(buf, sz, _TRUNCATE, fmt, arg)
```

being replaced by

```
#  define KRML_HOST_SNPRINTF(buf, sz, fmt, arg) _snprintf_s(buf, \
                                                            sz, \
                                                            _TRUNCATE, \
                                                            fmt, \
                                                            arg)
```

both of which are noticeably harder to read.

So: what maximum line length should we set? 80/90/95/100? Poll on Slack?

Reply:
I personally don't have strong feelings on this one personally, I'd be happy to use a longer line limit.

It's worth thinking about the other potential benefit of a short line limit: forcing people to write modular code that uses helper functions where appropriate. As the [Linux kernel coding style guide](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#indentation) puts it: "if you need more than 3 levels of indentation, you’re screwed anyway, and should fix your program". Having an 80-character line limit will in theory encourage better structure.

On the other hand Uncrustify won't help us to enforce this rule as it will only try to wrap long lines (it doesn't do hard line limit enforcement). We'd be asking people to restructure their code during review which we already do even without a hard line limit.

Reply:
To be clear, I'm not suggesting a longer line limit to enable deeply-nested arrow code. Rather, it's because over the years identifiers have got longer, and since functions take several parameters (from time to time) going beyond the 80-character limit is useful. Being required to split something that naturally reads as one line to cover three lines often hampers readability

Reply:
> These days, screens are much wider, and it's easy to have two windows with > 80-character lines next to each other, so the limit is readability rather than technology.

Technology and biology are still a limit. Not everyone has dual 30" monitors and a good enough eyesight for 8pt fonts.

80 characters has the advantage of being a standard. Having to readjust window sizes every time you switch between files sucks.

So I'm strongly in favor of keeping the guideline that you should be able to truncate the code at 80 columns, and the code should still be generally readable. Having `fier));` in the margin is ok but not great, having `fier + 3));` is not ok.

So having a tool that will stick to 80 columns where it's already decided to wrap, but won't rewrap lines just for the sake of fitting, is my preferred outcome.

Reply:
I use font size of 14 in VS Code, and a single monitor, so I'm on the more constrained side for line length, and I'm nonetheless in favour of longer line lengths than the 80 character limit here - from a quick look at the reformatted code, it looks overly narrow IMO.

Examples of wrapping like this:

```
    status = psa_verify_hash(key_id, psa_alg_md, hash, hash_len,
                             sig, sig_len);
```

just seem unneccessary and awkward to read to me.

Reply:
Linux and Zephyr have both changed their line length limit to 100 characters, with possible exceptions.

Reply:
My 2 pence here is that I am in favour of longer line limits, ideally 100 characters. I don't want to push people out though and so if there is a real risk of making people's lives harder/worse by extending the limit then I am happy to stick with 80, but my preference is for 100.

Reply:
I have a slight preference with 100 too, I'm fine smaller values (down to 80) as well, but I wouldn't be so happy with larger values.

(When working on my laptop, with the font size I'm using, my terminal window is just a bit more than 200 chars, so 100 allows me 1 vertical split, which I like.)

Reply:
My 2 pence here as well. My wish (dream?) about this is for the rule to be automatically enforced and thus do not spend time in code reviews on that at all. Thus I am fine with 100 especially if this allows us to move to automatic (and thus strict?) enforcement.

Reply:
Many good points have been raised. I think the arguments presented and the general consensus both favour a line length limit of 100 columns. With that in mind I'll set the limit to 100 unless future discussion overturns this decision.

Reply:
There's a majority but not a consensus here. I have a strong preference for a weak limit at 80.

I seem to be the only one with a preference for narrower lines, so I accept that I'm not going to have my way. But it _is_ going to make my life harder.

I've seen contradictory studies on line lengths, and generally hard to generalize: text and program code are different (and for both, language might matter), monitor technology matters, readers' eyesight matters…

(Yes, eyesight matters. Please don't make always this about monitor sizes. It's bad enough to see poor-shaming in many “stop living in the 20th century” calls for longer lines — but ok, those of us who are paid to work on Mbed TLS at Arm can have good monitors, though this does not necessarily apply to hobby contributors. But assuming that everyone gets the same number of readable columns out of X dollars of monitor is ableist.)

For me, around 60 columns is the maximum where I can read comfortably without shifting my eyes' focus horizontally. So code starting in column 8 (2 indents) and typically ending around column 70 or less is fine. Unimportant stuff like closing parentheses going a little beyond that is fine. Multiline comments where each line gets close to the right margin can be a little annoying sometimes. Having 100 columns as the usual right margin would make code, and especially comments, harder to read, regardless of the window width. (A smaller window and wrapping isn't a solution either: this doesn't break lines intelligently.)

To take Dave's example,
```
    status = psa_verify_hash(key_id, psa_alg_md, hash, hash_len,
                             sig, sig_len);
```
is fine to me, although I'd prefer
```
    status = psa_verify_hash(key_id, psa_alg_md,
                             hash, hash_len,
                             sig, sig_len);
```
so that the line is neatly divided into semantic clusters (configuration, hash, signature).
```
    status = psa_verify_hash(key_id, psa_alg_md, hash, hash_len, sig, sig_len);
```
fits in 80 columns, and I'd even write it myself, but I find it harder to parse with the long list of arguments. The reason I'd write this is to save screen real estate, because the worst thing you can do in terms of readability is to hide things off-screen.

Beyond visibility, there's also the issue of window size. It's far more convenient for me to have a single window size for everything, and 80 is both a long-established standard and one that's comfortable for most purposes. If I need to widen windows to 100 for Mbed TLS C code (and what about Python code?), I'd still want to wrap text at less than that, which means I have to start fiddling with window sizes when I move from file to file and I have to fit my windows on screen differently.

As for weak vs strong limit, I don't want a strong limit because I don't want people to have to break lines whenever they refactor code. If you give a variable a longer name and it puts a couple of lines into the margin, just let it go: adding line breaks is more work for the coder, more work for reviewers, more work positions across merges… Also, string literals should generally not be split.

I'd be ok with a hard limit at 100, but a soft limit at 80 (or even a little less): don't put meaningful stuff beyond ~72 and keep semicolons before ~78 when writing new code except when something (string literals, mostly) really can't be broken, but don't sweat refactorings, and at 100 the CI will fail.

Reply:
Good point that we have contributors (or even just readers) other that us, so it's not just about the personal preferences of team members.

I'd be happy about a soft limit at 80 and a hard at 100, but I'm not sure how that would be implemented. Would it be something like the style guide says "try to keep things at 80 most of the time" but the uncrustify config would just have 100. And reviewers are allowed/encouraged to complain when they see a line that goes over 80 for no good reason?

If so, I like the idea, but as Ronald noted, the drawback is that we loose some of the benefits of automatic enforcement. OTOH, not everything can be fully automated, for example the fact that when splitting long argument lists we should do it semantically (ie group `buf, buf_size, buf_len`), so perhaps we just have to accept that only 95% of style issues will be solved by enforcement and the remaining 5% will have to be handled manually. That would still be a huge progress - I mean, just solving issues of spaces around parenthesis would already be a huge progress.

On a somewhat related note, I'm a bit surprised that uncrustify had to break so many lines, I though we already had a limit a 80 chars (or a bit below, fwiw my editor is set to 78), apparently we were not very good and following that rule?